resources/assets/lib/utils/lang.ts

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import Lang from 'lang.js';
+import { formatNumber } from 'utils/html';
+import { present } from 'utils/string';
+
+type Replacement = string | number | null | undefined;
+
+export interface Replacements {
```

Comment:
it seems possible to just use the original typing by not setting nulls/undefined in the first place and formatting numbers beforehand (formatting them seems like a good idea anyway).

The most annoying one seems to be `formatTimestamp` but the function itself can probably be split to version that only accept number (if only conditional generic return is still supported in typescript...).

Reply:
actually, I think `formatTimestamp` might not need to handle `null` at all...